File: packages/shared-types/src/floor-plan-graph.ts
export interface Point {
  x: number;
  y: number;
}

export interface Wall {
  id: string;
  start: Point;
  end: Point;
  thickness: number;
  height: number;
  type: 'exterior' | 'interior';
}

export interface Room {
  id: string;
  label: string;
  polygon: Point[];
  level: number;
  area: number;
}

export interface Opening {
  id: string;
  type: 'door' | 'window' | 'sliding_door' | 'french_window';
  wallId: string;
  position: number;
  width: number;
  height: number;
  sillHeight: number;
  swingDirection?: 'left' | 'right' | 'inward' | 'outward';
}

export interface Stair {
  id: string;
  start: Point; // bottom of the staircase
  end: Point; // top of the staircase
  width: number; // mm
  numSteps: number;
  stepHeight: number; // mm (riser height)
  direction: 'up' | 'down';
  roomId?: string;
}

export interface FloorPlanGraph {
  version: 1;
  unit: 'mm';
  walls: Wall[];
  rooms: Room[];
  openings: Opening[];
  stairs: Stair[];
  metadata: {
    projectId: string;
    source: 'upload' | 'manual' | 'template';
    scale?: number;
    createdAt: string;
  };
}

File: services/parsing/app/schemas/floor_plan_graph.py
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel
from typing import Optional


class CamelModel(BaseModel):
    """Base model that serialises to camelCase for the TypeScript frontend."""
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )


class Point(BaseModel):
    x: float
    y: float


class WallSegment(CamelModel):
    id: str
    start: dict  # {x, y}
    end: dict    # {x, y}
    thickness: float = 10.0
    height: float = 270.0
    type: str = 'interior'


class RoomOutput(CamelModel):
    id: str
    label: str = ''
    polygon: list[dict]  # [{x, y}, ...]
    level: int = 0
    area: float = 0.0


class OpeningOutput(CamelModel):
    id: str
    type: str  # 'door' | 'window'
    wall_id: str
    position: float
    width: float
    height: float = 210.0
    sill_height: float = 0.0


class FloorPlanGraph(BaseModel):
    version: int = 1
    unit: str = 'mm'
    walls: list[dict]
    rooms: list[dict]
    openings: list[dict]
    confidence: float = 1.0
    metadata: dict = {}
